fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,147 @@
+use crate::{BlockHeight, Config, ParentFinalityProvider};
+use anyhow::{anyhow, Context};
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::ValidatorSet;
+use std::cmp::max;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer<T> {
+    config: Config,
+    started: Arc<AtomicBool>,
+    parent_view_provider: Arc<T>,
+    agent: Arc<IPCAgentProxy>,
+}
+
+impl<T> PollingParentSyncer<T> {
+    pub fn new(config: Config, parent_view_provider: Arc<T>, agent: Arc<IPCAgentProxy>) -> Self {
+        Self {
+            config,
+            started: Arc::new(AtomicBool::new(false)),
+            parent_view_provider,
+            agent,
+        }
+    }
+}
+
+impl<T: ParentFinalityProvider + Send + Sync + 'static> PollingParentSyncer<T> {
+    /// Start the proof of finality listener in the background
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let config = self.config.clone();
+        let provider = self.parent_view_provider.clone();
+        let agent = self.agent.clone();
+
+        tokio::spawn(async move {
+            if let Err(e) = sync_with_parent(config, agent, provider).await {
+                tracing::info!("sync with parent encountered error: {e}");
+            }
+        });
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: Config,
+    agent_proxy: Arc<IPCAgentProxy>,
+    provider: Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        // update block hash and validator set
+        let latest_height = agent_proxy
+            .get_chain_head_height()
+            .await
+            .context("cannot fetch parent chain head")?;
+
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+
+        let starting_height = if let Some(h) = provider.latest_height().await {
+            h + 1
+        } else {
+            // if latest_recorded_height is None, it means we have not started fetching any heights
+            // we just use the latest height minus a lower bound as the parent view
+            max(
+                1,
+                latest_height.saturating_sub(config.chain_head_lower_bound),
+            )
+        };
+
+        for h in starting_height..=latest_height {
+            let block_hash = agent_proxy
+                .get_block_hash(h)
+                .await
+                .context("cannot fetch block hash")?;
```

Comment:
What if there is no block at that height yet?

Reply:
it would raise an error and retry.